motor/discovery/azure/resolver.go

```diff
@@ -29,102 +31,123 @@ func (r *Resolver) AvailableDiscoveryTargets() []string {
 
 func (r *Resolver) Resolve(ctx context.Context, root *asset.Asset, tc *providers.Config, cfn common.CredentialFn, sfn common.QuerySecretFn, userIdDetectors ...providers.PlatformIdDetector) ([]*asset.Asset, error) {
 	resolved := []*asset.Asset{}
-
 	subscriptionID := tc.Options["subscription-id"]
+	clientId := tc.Options["client-id"]
 
-	// TODO: for now we only support the azure cli authentication
-	err := azure_provider.IsAzInstalled()
+	// Note: we use the resolver instead of the direct azure_provider.New to resolve credentials properly
+	m, err := resolver.NewMotorConnection(ctx, tc, cfn)
 	if err != nil {
 		return nil, err
 	}
-
-	// if we have no subscription, try to ask azure cli
-	if len(subscriptionID) == 0 {
-		log.Debug().Msg("no subscription id provided, fallback to azure cli")
-		// read from `az account show --output json`
-		account, err := azure_provider.GetAccount()
-		if err == nil {
-			subscriptionID = account.ID
-			// NOTE: we ignore the tenant id here since we validate it below
-		}
-		// if an error happens, the following config validation will catch the missing subscription id
+	defer m.Close()
+	provider, ok := m.Provider.(*azure_provider.Provider)
+	if !ok {
+		return nil, errors.New("could not create azure transport")
 	}
 
-	// Verify the subscription and get the details to ensure we have access
-	subscription, err := azure_provider.VerifySubscription(subscriptionID)
-	if err != nil || subscription.TenantID == nil {
-		return nil, errors.Wrap(err, "could not fetch azure subscription details for: "+subscriptionID)
+	// if no creds, check that the CLI is installed as we are going to use that
+	if clientId == "" && len(tc.Credentials) == 0 {
+		azInstalled := azure_provider.IsAzInstalled()
+		if !azInstalled {
+			return nil, errors.New("az not installed")
+		}
 	}
 
-	// attach tenant to config
-	tc.Options["tenant-id"] = *subscription.TenantID
-
-	provider, err := azure_provider.New(tc)
+	// get an authorizer to use for discovery
+	authorizer, err := provider.Authorizer()
 	if err != nil {
 		return nil, err
 	}
+	azureClient := NewAzureClient(authorizer)
 
-	identifier, err := provider.Identifier()
-	if err != nil {
-		return nil, err
+	// we either discover all subs (if no filter is provided) or just verify that the provided one exists
+	var subs []subscriptions.Subscription
+	subsClient := NewSubscriptions(azureClient)
+	if subscriptionID != "" {
+		subscription, err := subsClient.GetSubscription(subscriptionID)
+		if err != nil {
+			return nil, err
+		}
+		if tc.Options["tenant-id"] == "" {
+			tc.Options["tenant-id"] = *subscription.TenantID
```

Comment:
how does all this work with related assets/asset relationships? i may have missed it, but i didn't think i saw anywhere you're setting the related assets?

Reply:
It's indeed not there yet, we don't have anything about asset relationships in Azure so far. That can be a future PR, I don't know yet what we need to link together